pkg/timeutils/utils.go

```diff
@@ -1,36 +1,122 @@
+	RFC3339NanoLocal = "2006-01-02T15:04:05.999999999" // RFC3339Nano with local timezone
+	DateWithZone     = "2006-01-02Z07:00"              // RFC3339 with time at 00:00:00
+	DateLocal        = "2006-01-02"                    // RFC3339 with local timezone and time at 00:00:00
+)
+
 // GetTimestamp tries to parse given string as golang duration,
 // then RFC3339 time and finally as a Unix timestamp. If
 // any of these were successful, it returns a Unix timestamp
 // as string otherwise returns the given value back.
 // In case of duration input, the returned timestamp is computed
 // as the given reference time minus the amount of the duration.
-func GetTimestamp(value string, reference time.Time) string {
+func GetTimestamp(value string, reference time.Time) (string, error) {
 	if d, err := time.ParseDuration(value); value != "0" && err == nil {
-		return strconv.FormatInt(reference.Add(-d).Unix(), 10)
+		return strconv.FormatInt(reference.Add(-d).Unix(), 10), nil
 	}
 
 	var format string
+	var parseInLocation bool
+
+	// if the string has a Z or a + or three dashes use parse otherwise use parseinlocation
+	parseInLocation = !(strings.ContainsAny(value, "zZ+") || strings.Count(value, "-") == 3)
+
 	if strings.Contains(value, ".") {
-		format = time.RFC3339Nano
+		if parseInLocation {
+			format = RFC3339NanoLocal
+		} else {
+			format = time.RFC3339Nano
+		}
+	} else if strings.Contains(value, "T") {
+		// we want the number of colons in the T portion of the timestamp
+		tcolons := strings.Count(value, ":")
+		// if parseInLocation is off and we have a +/- zone offset (not Z) then
+		// there will be an extra colon in the input for the tz offset subract that
+		// colon from the tcolons count
+		if !parseInLocation && !strings.ContainsAny(value, "zZ") && tcolons > 0 {
+			tcolons--
+		}
+		if parseInLocation {
+			if tcolons >= 2 {
+				format = RFC3339Local
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04"
+			} else {
+				format = "2006-01-02T15"
+			}
+		} else {
+			if tcolons >= 2 {
+				format = time.RFC3339
+			} else if tcolons == 1 {
+				format = "2006-01-02T15:04Z07:00"
+			} else {
+				format = "2006-01-02T15Z07:00"
+			}
+		}
+	} else if parseInLocation {
+		format = DateLocal
+	} else {
+		format = DateWithZone
+	}
+
+	var t time.Time
+	var err error
+
+	if parseInLocation {
+		t, err = time.ParseInLocation(format, value, time.FixedZone(time.Now().Zone()))
 	} else {
-		format = time.RFC3339
+		t, err = time.Parse(format, value)
 	}
 
-	loc := time.FixedZone(time.Now().Zone())
-	if len(value) < len(format) {
-		format = format[:len(value)]
+	if err != nil {
+		// if there is a `-` then its an RFC3339 like timestamp otherwise assume unixtimestamp
+		if strings.Contains(value, "-") {
+			return "", err // was probably an RFC3339 like timestamp but the parser failed with an error
+		}
+		return value, nil // unixtimestamp in and out case
```

Comment:
it means we are returning what was passed into the function because we believe they entered seconds[.nanoseconds] from the command line.